src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/OnTypeFormattingHandler.cs

```diff
@@ -185,5 +209,53 @@ private static bool IsApplicableTriggerCharacter(string triggerCharacter, RazorL
             // Unknown trigger character.
             return false;
         }
+
+        private async Task<CSharpVirtualDocumentSnapshot> GetSynchronizedVirtualDocumentAsync(
+            Uri razorDocUri,
+            LSPDocumentSnapshot documentSnapshot,
+            CancellationToken cancellationToken)
+        {
+            if (!documentSnapshot.TryGetVirtualDocument(out CSharpVirtualDocumentSnapshot virtualDocumentSnapshot))
+            {
+                return null;
+            }
+
+            var synchronized = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync(
+                documentSnapshot.Version, virtualDocumentSnapshot, cancellationToken).ConfigureAwait(false);
+
+            LSPDocumentSnapshot updatedSnapshot = null;
+            if (synchronized && _documentManager.TryGetDocument(razorDocUri, out updatedSnapshot))
+            {
+                if (!updatedSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var updatedVirtualSnapshot))
+                {
+                    return null;
+                }
+
+                return updatedVirtualSnapshot;
+            }
+
+            // Recursively call this method until we reach a synchronized state. We need to be synchronized
+            // to ensure our C# virtual document is up to date.
+            return await GetSynchronizedVirtualDocumentAsync(razorDocUri, updatedSnapshot, cancellationToken).ConfigureAwait(false);
+        }
+
+        // Internal for testing
+        internal static Document GenerateRoslynCSharpDocument(SourceText csharpSourceText, VSHostServicesProvider hostServicesProvider)
+        {
+            var workspace = new AdhocWorkspace(hostServicesProvider.GetServices());
+            var project = workspace.AddProject("TestProject", LanguageNames.CSharp);
+            var document = workspace.AddDocument(project.Id, "TestDocument", csharpSourceText);
+            return document;
```

Comment:
I'm guessing if the tests pass, the answer is no, but do we not need the document options to be actually set on the document?

Also, should we dispose the AdhocWorkspace? I know the `FormattingContext` doesn't, but I've always thought that should be fixed, I just haven't got around to it.

Reply:
Ah, so the formatting methods take in an optional document option set which take priority over the document's options.

Good point about disposing the AdhocWorkspace, I updated the code. I _think_ the Dispose call is in the right place, I had to suppress the warning since we dispose in the caller after we're done using the document, but let me know if it looks off.

Reply:
I'll leave it up to you, but one other option would be to return a tuple (ie `var (workspace, document) = GenerateRoslynCSharpDocument(...)` just to make it a bit more explicit that the dispose is needed and remove the warning.

Though I'm not too fussed, because its not like its a public API (and I don't really care if the tests don't dispose anyway!)